File: src/main/java/com/vulpesnovis/StftFilter/Processor.java
package com.vulpesnovis.StftFilter;

import com.android.sdklib.util.SparseIntArray;
import edu.emory.mathcs.jtransforms.fft.DoubleFFT_1D;

import java.text.DecimalFormat;
import java.util.Arrays;
import java.util.Objects;

public class Processor {
    private int sampleRate; //inHz
    private short winSize; //in mS
    private int fftSize;
    private float winSizeInSamples;
    private final String func;

    //I'm introducing this block of code, because we cannot just pass fft result of entire audio file, to Output
    //It will be too much for an array, especially when audio file is relatively long
    //So, we need to split the entire fft result into several groups of windows

    private CompleteDataListener completeListener;
    private FunctionTestListener functionListener;
    private NewWindowListener windowListener;
    private boolean dbOutput;
    private int windowSize;

    /**
     *
     * @param fftSize number of samples (bins) in the output array, given to the power of two.
     * @param winSize width of the STFT window in mS.
     * @param dbOutput present output values normalized by dBFS.
     */
    public Processor(int sampleRate, int fftSize, int winSize, String func, CompleteDataListener listener, boolean dbOutput){
        this.sampleRate = sampleRate;
        this.func = Objects.requireNonNull(func);
        this.fftSize = (int)Math.pow(2,fftSize);
        this.winSize = (short)winSize;
        completeListener = Objects.requireNonNull(listener);
        this.dbOutput = dbOutput;
        winSizeInSamples = ((float)winSize/1000) / (1/(float)sampleRate);
    }
    public Processor(int sampleRate, int fftSize, String func, NewWindowListener listener, boolean dbOutput){
        this.sampleRate = sampleRate;
        this.func = Objects.requireNonNull(func);
        this.fftSize = (int)Math.pow(2,fftSize);
        windowListener = Objects.requireNonNull(listener);
        this.dbOutput = dbOutput;
    }
    public Processor(String func, FunctionTestListener listener, int windowSize, boolean dbOutput){
        this.func = Objects.requireNonNull(func);
        functionListener = Objects.requireNonNull(listener);
        this.windowSize = windowSize;
        this.dbOutput = dbOutput;
    }

    /**
     * JTransforms processing library is just brilliant
     * (but I don't think there's some kinda magic involved).
     * It somehow detects the Nyquist frequency of the source and makes the whole output calculation up to it.
     * Meanwhile, "n" argument of the constructor is our desired FFT size in bins.
     * And we just pass our windows one after another, to make a snapshot of the whole input array
     *
     * @param buffer array of decoded samples of the whole source file (or whatever we had in input)
     */
    public void process (double[] buffer){

        int winTotal = buffer.length / (int) winSizeInSamples;
        int samplesCount = 0; //total samples counter through all windows
        short timeOfInput = 0; //elapsed time in input signal
        float fftStep = (float)(sampleRate/2)/fftSize;

        //Keys starts from "00000000", including values at zero time and frequency
        SparseIntArray fftDataset = new SparseIntArray();
        //Time indexes starts from "0", from zero time value
        int[] arrayTimeValues = new int[winTotal];
        //Freq indexes starts from "1", excluding zero freq value, because it's not valid
        int[] arrayFreqValues = new int[fftSize];
        double magMax = -170;
        double magMin = 0;

        for (int i = skipIdle(buffer); i < winTotal; i++) {//cycle for each window in the whole output sequence

            DoubleFFT_1D transformer = new DoubleFFT_1D(fftSize*2);
            double[] window = new double [(int)winSizeInSamples];
            double[] windowNResult = new double[Integer.max(fftSize * 2, (int) winSizeInSamples)];
            System.arraycopy(buffer, samplesCount, window, 0, (int)winSizeInSamples);

            System.arraycopy(Windows.applyWindow(window, func), 0, windowNResult, 0, window.length);

            transformer.realForward(windowNResult);
//            transformer.realForward(windowNResult);
            float j = fftStep; //iterator for frequency val in output array (key)

            //cycle for each bin in every single window transform.
            boolean replaceBroken = false;
            for (int k = 1; k < fftSize; k++) {
                double re = windowNResult[2*k];
                double im = windowNResult[2*k+1];
                double mag = Math.sqrt(re*re + im*im);

                if (dbOutput){//writing array in absolute values
                    double ref = 470; //dBFS reference value measured on pure sine wave with no window function applied
                    mag = 20 * Math.log10(mag/ref);
                }
                //These two values coded into one key
                //Number of bin consists of 4 digits, and number of windows is 4 digits either (let's assume it's our input file limitation)

                DecimalFormat df = new DecimalFormat("0000");
                String timeIndex = df.format(i);
                String freqIndex = df.format(k);
                String key = timeIndex + freqIndex;

                arrayFreqValues[k] = (int)(j*100);
                if (re==0&im==0) mag=magMin;
                if (magMax < mag & mag != 0) magMax = mag;
                if (magMin > mag) magMin=mag;

                if (replaceBroken & k>1){
                    String memKey = df.format(i) + df.format(k-2);
                    String prevKey = df.format(i) + df.format(k-1);
                    int prevMag = (int)((fftDataset.get(Integer.parseInt(memKey))+mag)/2);
                    fftDataset.put(Integer.parseInt(prevKey), prevMag);
                    if (i==0) System.out.println("\u001B[34mFreaking hell. Value approximated again. Key: " + df.format(k-1) + ". Something wrong with ur input" + "\u001B[0m");
                    replaceBroken = false;
                }
                replaceBroken = re == 0 & im == 0;

                fftDataset.append(Integer.parseInt(key), (int)mag);

                j+=fftStep;
            }

            arrayTimeValues[i] = timeOfInput;
            timeOfInput += winSize;
            samplesCount += winSizeInSamples;
        }
        completeListener.onDataComputed(arrayTimeValues, arrayFreqValues, fftDataset, (int)magMin, (int)magMax);
    }
    public void processSingle(double[] buffer){
        if (buffer[0]==0){
            windowListener.onIdlePassed();
            return;
        }

        int winSizeInSamples = buffer.length;
        float fftStep = (float)(sampleRate/2)/fftSize;

        SparseIntArray fftSnapshot = new SparseIntArray();
        double magMax = -170;
        double magMin = 0;


        DoubleFFT_1D transformer = new DoubleFFT_1D(fftSize*2);
        double[] windowNResult = new double[Integer.max(fftSize * 2, winSizeInSamples)];

        System.arraycopy(Windows.applyWindow(buffer, func), 0, windowNResult, 0, buffer.length);

        transformer.realForward(windowNResult);
        float j = fftStep; //iterator for frequency val in output array (key)

        //cycle for each bin in every single window transform.
        boolean replaceBroken = false;
        for (int k = 1; k < fftSize; k++) {
            double re = windowNResult[2*k];
            double im = windowNResult[2*k+1];
            double mag = Math.sqrt(re*re + im*im);

            if (dbOutput){//writing array in absolute values
                double ref = 470; //dBFS reference value measured on pure sine wave with no window function applied
                mag = 20 * Math.log10(mag/ref);
            }

            int key = (int)(j*100);

            if (re==0&im==0) mag=magMin;
            if (magMax < mag & mag != 0) magMax = mag;
            if (magMin > mag) magMin=mag;

            if (replaceBroken & k>1){
                int memKey = (int)((j-2*fftStep)*100);
                int prevKey = (int)((j-fftStep)*100);
                int prevMag = (int)((fftSnapshot.get(memKey)+mag)/2);
                fftSnapshot.put(prevKey, prevMag);
                System.out.println("\u001B[34mFreaking hell. Value approximated again. Freq: " + prevKey + ". Something wrong with ur input" + "\u001B[0m");
                replaceBroken = false;
            }
            replaceBroken = re == 0 & im == 0;

            fftSnapshot.append(key, (int)mag);

            j+=fftStep;
        }

        fftSnapshot.put(0,0);
        windowListener.onWindowComputed(fftSnapshot);
    }
    public void processFunction (){
        double[] window = new double[windowSize];
        Arrays.fill(window, 1);

        functionListener.onRawFunctionComputed(Windows.applyWindow(window, func));
    }
    //This method takes control over the data processing, to skip the initial amount of zero values of the input dataset.
    //And returns value for the window counter to start real processing
    private int skipIdle(double[] buffer){
        int start=0;
        for (int i = 0; i < buffer.length; i++) {
            if (buffer[i]!=0){
                start = Math.ceilDiv(i, (int)winSizeInSamples);
                break;
            }
        }
        if (start==0)
            throw new NullPointerException("Input buffer is null");
        return start-1;
    }
}

File: src/main/java/com/vulpesnovis/StftFilter/Processor.kt
package com.vulpesnovis.StftFilter

import com.android.sdklib.util.SparseIntArray
import edu.emory.mathcs.jtransforms.fft.DoubleFFT_1D
import java.text.DecimalFormat
import java.util.*
import kotlin.math.log10
import kotlin.math.pow
import kotlin.math.sqrt

class Processor private constructor(){
    private var sampleRate = 0
    private var winSize = 20
    private var fftSize = 256
    private lateinit var func: String
    private var winSizeInSamples = 0f
    private lateinit var processingFun: (DoubleArray) -> Unit

    //I'm introducing this block of code, because we cannot just pass fft result of entire audio file, to Output
    //It will be too much for an array, especially when audio file is relatively long
    //So, we need to split the entire fft result into several groups of windows
    private lateinit var completeListener: CompleteDataListener
    private lateinit var functionListener: FunctionTestListener
    private lateinit var windowListener: NewWindowListener
    private var dbOutput = false
    private var windowSize = 0


    constructor(sampleRate: Int, fftSize: Int, winSize: Int, func: String, listener: CompleteDataListener, dbOutput: Boolean) : this(){
        this.sampleRate = sampleRate
        this.func = func
        this.fftSize = 2.0.pow(fftSize.toDouble()).toInt()
        this.winSize = winSize.toShort().toInt()
        this.winSizeInSamples = winSize.toFloat() / 1000 / (1 / sampleRate.toFloat())
        this.dbOutput = dbOutput
        completeListener = listener
        processingFun = ::processComplete
    }
    constructor(sampleRate: Int, fftSize: Int, func: String, listener: NewWindowListener, dbOutput: Boolean) : this(){
        this.sampleRate = sampleRate
        this.func = func
        this.fftSize = 2.0.pow(fftSize.toDouble()).toInt()
        this.dbOutput = dbOutput
        windowListener = listener
        processingFun = ::processComplete
    }
    constructor(func: String, listener: FunctionTestListener, windowSize: Int, dbOutput: Boolean) : this(){
        this.func = func
        this.windowSize = windowSize
        this.dbOutput = dbOutput
        functionListener = listener
        processingFun = ::processFunction
    }

    fun process(buffer: DoubleArray) = processingFun(buffer)

    private fun processComplete(buffer: DoubleArray){
        val winTotal = buffer.size / winSizeInSamples.toInt()
        var samplesCount = 0 //total samples counter through all windows

        var timeOfInput: Int = 0 //elapsed time in input signal

        val fftStep = (sampleRate / 2).toFloat() / fftSize

        //Keys starts from "00000000", including values at zero time and frequency
        val fftDataset = SparseIntArray()
        //Time indexes starts from "0", from zero time value
        val arrayTimeValues = IntArray(winTotal)
        //Freq indexes starts from "1", excluding zero freq value, because it's not valid
        val arrayFreqValues = IntArray(fftSize)
        var magMax = -170.0
        var magMin = 0.0

        for (i in skipIdle(buffer) until winTotal) { //cycle for each window in the whole output sequence
            val transformer = DoubleFFT_1D(fftSize * 2)
            val window = DoubleArray(winSizeInSamples.toInt())
            val windowNResult = DoubleArray(Integer.max(fftSize * 2, winSizeInSamples.toInt()))
            System.arraycopy(buffer, samplesCount, window, 0, winSizeInSamples.toInt())
            System.arraycopy(Windows.applyWindow(window, func), 0, windowNResult, 0, window.size)
            transformer.realForward(windowNResult)
            //            transformer.realForward(windowNResult);
            var j = fftStep //iterator for frequency val in output array (key)

            //cycle for each bin in every single window transform.
            var replaceBroken = false
            for (k in 1 until fftSize) {
                val re = windowNResult[2 * k]
                val im = windowNResult[2 * k + 1]
                var mag = sqrt(re * re + im * im)
                if (dbOutput) { //writing array in absolute values
                    val ref = 470.0 //dBFS reference value measured on pure sine wave with no window function applied
                    mag = 20 * log10(mag / ref)
                }
                //These two values coded into one key
                //Number of bin consists of 4 digits, and number of windows is 4 digits either (let's assume it's our input file limitation)
                val df = DecimalFormat("0000")
                val timeIndex: String = df.format(i.toLong())
                val freqIndex = df.format(k.toLong())
                val key = timeIndex + freqIndex
                arrayFreqValues[k] = (j * 100).toInt()
                if ((re == 0.0) and (im == 0.0)) mag = magMin
                if ((magMax < mag) and (mag != 0.0)) magMax = mag
                if (magMin > mag) magMin = mag
                if (replaceBroken and (k > 1)) {
                    val memKey: String = df.format(i.toLong()) + df.format((k - 2).toLong())
                    val prevKey: String = df.format(i.toLong()) + df.format((k - 1).toLong())
                    val prevMag = ((fftDataset[memKey.toInt()] + mag) / 2).toInt()
                    fftDataset.put(prevKey.toInt(), prevMag)
                    if (i == 0) println("\u001B[34mFreaking hell. Value approximated again. Key: " + df.format((k - 1).toLong()) + ". Something wrong with ur input" + "\u001B[0m")
                    replaceBroken = false
                }
                replaceBroken = (re == 0.0) and (im == 0.0)
                fftDataset.append(key.toInt(), mag.toInt())
                j += fftStep
            }
            arrayTimeValues[i] = timeOfInput.toInt()
            timeOfInput += winSize
            samplesCount += winSizeInSamples.toInt()
        }
        completeListener.onDataComputed(arrayTimeValues, arrayFreqValues, fftDataset, magMin.toInt(), magMax.toInt())

    }
    private fun skipIdle(buffer: DoubleArray): Int {
        var start = 0
        for (i in buffer.indices) {
            if (buffer[i] != 0.0) {
                start = Math.ceilDiv(i, winSizeInSamples.toInt())
                break
            }
        }
        if (start == 0) throw NullPointerException("Input buffer is null")
        return start - 1
    }

    fun processSingle(buffer: DoubleArray) {
        if (buffer[0] == 0.0) {
            windowListener.onIdlePassed()
            return
        }
        val winSizeInSamples = buffer.size
        val fftStep = (sampleRate / 2).toFloat() / fftSize
        val fftSnapshot = SparseIntArray()
        var magMax = -170.0
        var magMin = 0.0
        val transformer = DoubleFFT_1D(fftSize * 2)
        val windowNResult = DoubleArray(Integer.max(fftSize * 2, winSizeInSamples))
        System.arraycopy(Windows.applyWindow(buffer, func), 0, windowNResult, 0, buffer.size)
        transformer.realForward(windowNResult)
        var j = fftStep //iterator for frequency val in output array (key)

        //cycle for each bin in every single window transform.
        var replaceBroken = false
        for (k in 1 until fftSize) {
            val re = windowNResult[2 * k]
            val im = windowNResult[2 * k + 1]
            var mag = sqrt(re * re + im * im)
            if (dbOutput) { //writing array in absolute values
                val ref = 470.0 //dBFS reference value measured on pure sine wave with no window function applied
                mag = 20 * log10(mag / ref)
            }
            val key = (j * 100).toInt()
            if ((re == 0.0) and (im == 0.0)) mag = magMin
            if ((magMax < mag) and (mag != 0.0)) magMax = mag
            if (magMin > mag) magMin = mag
            if (replaceBroken and (k > 1)) {
                val memKey = ((j - 2 * fftStep) * 100).toInt()
                val prevKey = ((j - fftStep) * 100).toInt()
                val prevMag = ((fftSnapshot[memKey] + mag) / 2).toInt()
                fftSnapshot.put(prevKey, prevMag)
                println("\u001B[34mFreaking hell. Value approximated again. Freq: $prevKey. Something wrong with ur input\u001B[0m")
                replaceBroken = false
            }
            replaceBroken = (re == 0.0) and (im == 0.0)
            fftSnapshot.append(key, mag.toInt())
            j += fftStep
        }
        fftSnapshot.put(0, 0)
        windowListener.onWindowComputed(fftSnapshot)
    }

    private fun processFunction(dummy: DoubleArray) {
        val window = DoubleArray(windowSize)
        Arrays.fill(window, 1.0)
        functionListener.onRawFunctionComputed(Windows.applyWindow(window, func))
    }
}
